File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/tasks/TaskReporter.java
package ru.kmorozov.onedrive.tasks;

import ru.kmorozov.onedrive.client.utils.LogUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class TaskReporter {

    private static final Logger log = LogManager.getLogger(TaskReporter.class.getName());

    private int same;
    private int remoteDeleted;
    private int localDeleted;
    private int skipped;
    private int propsUpdated;
    private int errors;

    private int newUploaded;
    private long newUploadedSize;
    private int replaceUploaded;
    private long replaceUploadedSize;

    private int newDownloaded;
    private long newDownloadedSize;
    private int replaceDownloaded;
    private long replaceDownloadedSize;

    private Logger taskLogger;

    private final long startTime;

    public TaskReporter() {
        this.startTime = System.currentTimeMillis();
    }

    public synchronized void same() {
        this.same++;
    }

    public synchronized void remoteDeleted() {
        this.remoteDeleted++;
    }

    public synchronized void localDeleted() {
        this.localDeleted++;
    }

    public synchronized void skipped() {
        this.skipped++;
    }

    public synchronized void error() {
        this.errors++;
    }

    public synchronized void fileUploaded(boolean replace, long size) {
        if (replace) {
            this.replaceUploaded++;
            this.replaceUploadedSize += size;
        } else {
            this.newUploaded++;
            this.newUploadedSize += size;
        }
    }

    public synchronized void fileDownloaded(boolean replace, long size) {
        if (replace) {
            this.replaceDownloaded++;
            this.replaceDownloadedSize += size;
        } else {
            this.newDownloaded++;
            this.newDownloadedSize += size;
        }
    }

    public synchronized void propertiesUpdated() {
        this.propsUpdated++;
    }

    public synchronized void report() {

        if (0 < this.errors) {
            TaskReporter.log.error(String.format("%d tasks failed - see log for details", this.errors));
        }

        if (0 < this.same) {
            TaskReporter.log.info(String.format("Skipped %d unchanged file%s", this.same, TaskReporter.plural((long) this.same)));
        }

        if (0 < this.skipped) {
            TaskReporter.log.info(String.format("Skipped %d ignored file%s", this.skipped, TaskReporter.plural((long) this.skipped)));
        }

        if (0 < this.localDeleted) {
            TaskReporter.log.info(String.format("Deleted %d local file%s", this.localDeleted, TaskReporter.plural((long) this.skipped)));
        }

        if (0 < this.remoteDeleted) {
            TaskReporter.log.info(String.format("Deleted %d remote file%s", this.remoteDeleted, TaskReporter.plural((long) this.skipped)));
        }

        if (0 < this.propsUpdated) {
            TaskReporter.log.info(String.format("Updated timestamps on %d file%s", this.propsUpdated, TaskReporter.plural((long) this.skipped)));
        }

        if (0 < this.newUploaded || 0 < this.replaceUploaded) {

            StringBuilder uploadedResult = new StringBuilder();

            uploadedResult.append(
                    String.format("Uploaded %d file%s (%s) - ",
                            this.newUploaded + this.replaceUploaded,
                            TaskReporter.plural((long) (this.newUploaded + this.replaceUploaded)),
                                  LogUtils.readableFileSize(this.newUploadedSize + this.replaceUploadedSize)));

            if (0 < this.newUploaded) {
                uploadedResult.append(
                        String.format("%d new file%s (%s) ",
                                this.newUploaded,
                                TaskReporter.plural((long) this.newUploaded),
                                      LogUtils.readableFileSize(this.newUploadedSize)));
            }

            if (0 < this.replaceUploaded) {
                uploadedResult.append(
                        String.format("%d new file%s (%s) ",
                                this.replaceUploaded,
                                TaskReporter.plural((long) this.replaceUploaded),
                                      LogUtils.readableFileSize(this.replaceUploadedSize)));
            }

            TaskReporter.log.info(uploadedResult.toString());
        }

        if (0 < this.newDownloaded || 0 < this.replaceDownloaded) {
            StringBuilder downloadedResult = new StringBuilder();

            downloadedResult.append(
                    String.format("Downloaded %d file%s (%s) - ",
                            this.newDownloaded + this.replaceDownloaded,
                            TaskReporter.plural((long) (this.newDownloaded + this.replaceDownloaded)),
                                  LogUtils.readableFileSize(this.newDownloadedSize + this.replaceDownloadedSize)));

            if (0 < this.newDownloaded) {
                downloadedResult.append(
                        String.format("%d new file%s (%s) ",
                                this.newDownloaded,
                                TaskReporter.plural((long) this.newDownloaded),
                                      LogUtils.readableFileSize(this.newDownloadedSize)));
            }

            if (0 < this.replaceDownloaded) {
                downloadedResult.append(
                        String.format("%d new file%s (%s) ",
                                this.replaceDownloaded,
                                TaskReporter.plural((long) this.replaceDownloaded),
                                      LogUtils.readableFileSize(this.replaceDownloadedSize)));
            }

            TaskReporter.log.info(downloadedResult.toString());
        }

        long elapsed = System.currentTimeMillis() - this.startTime;
        TaskReporter.log.info(String.format("Elapsed time: %s", LogUtils.readableTime(elapsed)));
    }

    private static String plural(long same) {
        return 1L == same ? "" : "s";
    }

    public void setTaskLogger(Logger taskLogger) {
        this.taskLogger = taskLogger;
    }

    public void info(String message) {
        if (null != this.taskLogger)
            this.taskLogger.info(message);
        else
            TaskReporter.log.info(message);
    }

    public void warn(String message) {
        if (null != this.taskLogger)
            this.taskLogger.warn(message);
        else
            TaskReporter.log.warn(message);
    }
}

File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/tasks/TaskReporter.kt
package ru.kmorozov.onedrive.tasks

import ru.kmorozov.onedrive.client.utils.LogUtils
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

open class TaskReporter {

    private var same: Int = 0
    private var remoteDeleted: Int = 0
    private var localDeleted: Int = 0
    private var skipped: Int = 0
    private var propsUpdated: Int = 0
    private var errors: Int = 0

    private var newUploaded: Int = 0
    private var newUploadedSize: Long = 0
    private var replaceUploaded: Int = 0
    private var replaceUploadedSize: Long = 0

    private var newDownloaded: Int = 0
    private var newDownloadedSize: Long = 0
    private var replaceDownloaded: Int = 0
    private var replaceDownloadedSize: Long = 0

    private var taskLogger: Logger? = null

    private val startTime: Long

    init {
        startTime = System.currentTimeMillis()
    }

    @Synchronized
    fun same() {
        same++
    }

    @Synchronized
    fun remoteDeleted() {
        remoteDeleted++
    }

    @Synchronized
    fun localDeleted() {
        localDeleted++
    }

    @Synchronized
    fun skipped() {
        skipped++
    }

    @Synchronized
    fun error() {
        errors++
    }

    @Synchronized
    fun fileUploaded(replace: Boolean, size: Long) {
        if (replace) {
            replaceUploaded++
            replaceUploadedSize += size
        } else {
            newUploaded++
            newUploadedSize += size
        }
    }

    @Synchronized
    fun fileDownloaded(replace: Boolean, size: Long) {
        if (replace) {
            replaceDownloaded++
            replaceDownloadedSize += size
        } else {
            newDownloaded++
            newDownloadedSize += size
        }
    }

    @Synchronized
    fun propertiesUpdated() {
        propsUpdated++
    }

    @Synchronized
    fun report() {

        if (0 < errors) {
            log.error(String.format("%d tasks failed - see log for details", errors))
        }

        if (0 < same) {
            log.info(String.format("Skipped %d unchanged file%s", same, plural(same.toLong())))
        }

        if (0 < skipped) {
            log.info(String.format("Skipped %d ignored file%s", skipped, plural(skipped.toLong())))
        }

        if (0 < localDeleted) {
            log.info(String.format("Deleted %d local file%s", localDeleted, plural(skipped.toLong())))
        }

        if (0 < remoteDeleted) {
            log.info(String.format("Deleted %d remote file%s", remoteDeleted, plural(skipped.toLong())))
        }

        if (0 < propsUpdated) {
            log.info(String.format("Updated timestamps on %d file%s", propsUpdated, plural(skipped.toLong())))
        }

        if (0 < newUploaded || 0 < replaceUploaded) {

            val uploadedResult = StringBuilder()

            uploadedResult.append(
                    String.format("Uploaded %d file%s (%s) - ",
                            newUploaded + replaceUploaded,
                            plural((newUploaded + replaceUploaded).toLong()),
                            LogUtils.readableFileSize(newUploadedSize + replaceUploadedSize)))

            if (0 < newUploaded) {
                uploadedResult.append(
                        String.format("%d new file%s (%s) ",
                                newUploaded,
                                plural(newUploaded.toLong()),
                                LogUtils.readableFileSize(newUploadedSize)))
            }

            if (0 < replaceUploaded) {
                uploadedResult.append(
                        String.format("%d new file%s (%s) ",
                                replaceUploaded,
                                plural(replaceUploaded.toLong()),
                                LogUtils.readableFileSize(replaceUploadedSize)))
            }

            log.info(uploadedResult.toString())
        }

        if (0 < newDownloaded || 0 < replaceDownloaded) {
            val downloadedResult = StringBuilder()

            downloadedResult.append(
                    String.format("Downloaded %d file%s (%s) - ",
                            newDownloaded + replaceDownloaded,
                            plural((newDownloaded + replaceDownloaded).toLong()),
                            LogUtils.readableFileSize(newDownloadedSize + replaceDownloadedSize)))

            if (0 < newDownloaded) {
                downloadedResult.append(
                        String.format("%d new file%s (%s) ",
                                newDownloaded,
                                plural(newDownloaded.toLong()),
                                LogUtils.readableFileSize(newDownloadedSize)))
            }

            if (0 < replaceDownloaded) {
                downloadedResult.append(
                        String.format("%d new file%s (%s) ",
                                replaceDownloaded,
                                plural(replaceDownloaded.toLong()),
                                LogUtils.readableFileSize(replaceDownloadedSize)))
            }

            log.info(downloadedResult.toString())
        }

        val elapsed = System.currentTimeMillis() - startTime
        log.info(String.format("Elapsed time: %s", LogUtils.readableTime(elapsed)))
    }

    fun setTaskLogger(taskLogger: Logger) {
        this.taskLogger = taskLogger
    }

    open fun info(message: String) {
        if (null != taskLogger)
            taskLogger!!.info(message)
        else
            log.info(message)
    }

    open fun warn(message: String) {
        if (null != taskLogger)
            taskLogger!!.warn(message)
        else
            log.warn(message)
    }

    companion object {

        private val log = LogManager.getLogger(TaskReporter::class.java.name)

        private fun plural(same: Long): String {
            return if (1L == same) "" else "s"
        }
    }
}
